{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "## @package dna2aa \n",
    "#   Translate the input sequence in the specified or all frames.\n",
    "#\n",
    "#   Output: .fa file of translated subject sequences.\n",
    "#   These sequences are naively translated as internal ORFs,\n",
    "#   and are likely larger than the sequence's true ORF.\n",
    "#   The output is best suited to use for hmmer-like alignments.\n",
    "#\n",
    "#   Alternatively, using the 'seqbreak' option, ORFs are broken\n",
    "#   at stop codons.\n",
    "#\n",
    "#   Written by Austin Hammond, GSC 2015\n",
    "\n",
    "import argparse\n",
    "import sys\n",
    "import csv\n",
    "\n",
    "import fastatools as ft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mtdna2aa(seq):\n",
    "    tbl = {\"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"M\", \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "        \"TTA\":\"L\", \"TTG\":\"L\", \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\", \"TTT\":\"F\",\n",
    "        \"TTC\":\"F\", \"ATG\":\"M\", \"TGT\":\"C\", \"TGC\":\"C\", \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\",\n",
    "        \"GCG\":\"A\", \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\", \"CCT\":\"P\", \"CCC\":\"P\",\n",
    "        \"CCA\":\"P\", \"CCG\":\"P\", \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\", \"TCT\":\"S\",\n",
    "        \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\", \"TAT\":\"Y\", \"TAC\":\"Y\",\n",
    "        \"TGG\":\"W\", \"CAA\":\"Q\", \"CAG\":\"Q\", \"AAT\":\"N\", \"AAC\":\"N\", \"CAT\":\"H\", \"CAC\":\"H\",\n",
    "        \"GAA\":\"E\", \"GAG\":\"E\", \"GAT\":\"D\", \"GAC\":\"D\", \"AAA\":\"K\", \"AAG\":\"K\", \"CGT\":\"R\",\n",
    "        \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\", \"AGA\":\"-\", \"AGG\":\"-\", \"TAA\":\"-\", \"TAG\":\"-\",\n",
    "        \"TGA\":\"W\"}\n",
    "    j = 0 # for iteration\n",
    "    aa = ''\n",
    "    seql = len(seq)\n",
    "    if seql % 3 != 0:\n",
    "        seq = seq[0:(seql-seql%3)] #trim to multiple of 3, but should be already\n",
    "    seql = len(seq) #recalculate\n",
    "    for j in range(0,seql,3):\n",
    "        j += 1\n",
    "        if j == (seql-1):\n",
    "            codon = seq[j:]\n",
    "        else:\n",
    "            codon = seq[j-1:j+2]\n",
    "    # print codon\n",
    "        aa = aa + tbl[codon]\n",
    "    # force first codon to be M\n",
    "    aa = \"M\" + aa[1:]\n",
    "\n",
    "    return aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mtsixframe(seq):\n",
    "    one = mtdna2aa(seq)\n",
    "    two = mtdna2aa(seq[1:])\n",
    "    three = mtdna2aa(seq[2:])\n",
    "    none = mtdna2aa(ft.revcomp(seq))\n",
    "    ntwo = mtdna2aa(ft.revcomp(seq)[1:])\n",
    "    nthree = mtdna2aa(ft.revcomp(seq)[2:])\n",
    "\n",
    "    return [one, two, three, none, ntwo, nthree]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup parser\n",
    "parser = argparse.ArgumentParser(\n",
    "    description='Naively translate the input DNA sequence.')\n",
    "parser.add_argument(\n",
    "    '--sequence', '-s', action='store', help=\"The fasta file of DNA sequences [STDIN]\",\n",
    "    default='stdin')\n",
    "parser.add_argument('--frame', '-f', action='store',\n",
    "    choices=['+1','+2','+3','-1','-2','-3','all'],\n",
    "    help=\"Frame to use for sequence translation [all]\", default='all')\n",
    "parser.add_argument('--seqbreak','-b',action='store_true',\n",
    "    help=\"Break translations at in-frame stop codons\")\n",
    "parser.add_argument('--outfile', '-o', action='store',\n",
    "    help='Name to use for output fasta file [STDOUT]', default='stdout')\n",
    "parser.add_argument('--mito', '-m', action='store_true', \n",
    "    help=('Sequence is mitochondrial; use vertebrate Mt code.'\n",
    "        ' Does not allow stop codons to be completed by polyadenylation.'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "### FOR DEV ###\n",
    "#frm = 'all'\n",
    "outfile = 'stdout'\n",
    "#outfile = 'tests/test-out.txt'\n",
    "fasta = 'tests/untitled.txt'\n",
    "fastadic = {}\n",
    "outdic = {}\n",
    "seqbreak = True\n",
    "mito = False\n",
    "frm = '+1'\n",
    "### END DEV ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse args\n",
    "args = parser.parse_args()\n",
    "frm = args.frame\n",
    "fasta = args.sequence\n",
    "fastadic = {}\n",
    "outfile = args.outfile\n",
    "outdic = {}\n",
    "mito = args.mito\n",
    "seqbreak = args.seqbreak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check that the frame was passed properly\n",
    "if frm == 'all':\n",
    "    pass\n",
    "elif \"+\" in frm:\n",
    "    pass\n",
    "elif frm == '1':\n",
    "    frm = '+1'\n",
    "elif frm == '2':\n",
    "    frm = '+2'\n",
    "elif frm == '3':\n",
    "    frm = '+3'\n",
    "elif \"-\" not in frm:\n",
    "    print \"ERROR: I didn't understand what reading frame you want.\"\n",
    "    print \"Please specify 'all' or a single frame including + or - and retry.\"\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define possible reading frames\n",
    "if frm == 'all':\n",
    "    frames = ['+1','+2','+3','-1','-2','-3']\n",
    "else:\n",
    "    frames = [frm]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in sequence input\n",
    "if fasta == 'stdin':\n",
    "    for rec in ft.fasta_iter(sys.stdin):\n",
    "        fastadic[rec[0].split(\" \")[0]] = rec[1]\n",
    "\n",
    "else:\n",
    "    with open(fasta) as fa:\n",
    "        for rec in ft.fasta_iter(fa):\n",
    "            sid = rec[0].split(\" \")[0]\n",
    "            sqn = rec[1]\n",
    "            fastadic[sid] = sqn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "if seqbreak:\n",
    "    # setup separate dict to hold split translations\n",
    "    transdic = {}\n",
    "\n",
    "for sid, sqn in fastadic.items():\n",
    "    if mito == True:\n",
    "        if frm == 'all':\n",
    "            trns = mtsixframe(sqn)\n",
    "        elif frm == '+1':\n",
    "            trns = mtdna2aa(sqn)\n",
    "        elif frm == '+2':\n",
    "            trns = mtdna2aa(sqn[1:])\n",
    "        elif frm == '+3':\n",
    "            trns = mtdna2aa(sqn[2:])\n",
    "        elif frm == '-1':\n",
    "            trns = mtdna2aa(ft.revcomp(sqn))\n",
    "        elif frm == '-2':\n",
    "            trns = mtdna2aa(ft.revcomp(sqn)[1:])\n",
    "        else:\n",
    "            trns = mtdna2aa(ft.revcomp(sqn)[2:])\n",
    "\n",
    "    else:\n",
    "        if frm == 'all':\n",
    "            trns = ft.sixframe(sqn)\n",
    "        elif frm == '+1':\n",
    "            trns = ft.dna2aa(sqn)\n",
    "        elif frm == '+2':\n",
    "            trns = ft.dna2aa(sqn[1:])\n",
    "        elif frm == '+3':\n",
    "            trns = ft.dna2aa(sqn[2:])\n",
    "        elif frm == '-1':\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn))\n",
    "        elif frm == '-2':\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn)[1:])\n",
    "        else:\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn)[2:])\n",
    "\n",
    "    # setup output dict entry\n",
    "    outdic[sid] = {}\n",
    "    if seqbreak:\n",
    "        # if frm == all, trns is a list; otherwise is a string #TODO make always a list\n",
    "        if type(trns) == list:\n",
    "            if frm == 'all':\n",
    "                transzip = zip(frames, trns)\n",
    "            else:\n",
    "                transzip = (frm, trns)\n",
    "            for rec in transzip:\n",
    "                thisframe = rec[0]\n",
    "                thisseq = rec[1]\n",
    "                #splitsies = [x.split(\"-\") for x in thisseq]\n",
    "                splitsies = thisseq.split(\"-\")\n",
    "                # drop empty frames\n",
    "                splitsies = [x for x in splitsies if x]\n",
    "                outdic[sid][thisframe] = splitsies\n",
    "        else:\n",
    "            splitsies = trns.split(\"-\") # split into ORFs\n",
    "            multi = [x for x in splitsies if x] # drop empty ORFs\n",
    "            outdic[sid][frm] = multi\n",
    "    else:\n",
    "        if type(trns) == list:\n",
    "            outdic[sid] = {}\n",
    "            if frm == 'all':\n",
    "                transzip = zip(frames, trns)\n",
    "            else:\n",
    "                transzip = (frm, trns)\n",
    "            for rec in transzip:\n",
    "                thisframe = rec[0]\n",
    "                thisseq = rec[1]\n",
    "                outdic[sid][thisframe] = thisseq\n",
    "        else:\n",
    "            outdic[sid][frm] = trns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "frm +1\n",
      "frames ['+1']\n",
      "fastadic {'seq3': 'TGGATGACACTAGGCATATAA', 'seq2': 'TGGATTACACTAGGCATATAA', 'seq1': 'ATGGATTACACTAGGCATTAA', 'seq5': 'ATGGAACATTACTAATAATAA', 'seq4': 'TGGATGACACTAGGCATATAAAAT'}\n",
      "trns WMTLGI-N\n",
      "transzip [('+1', 'WMTLGI-N'), ('+2', 'G-H-AYK'), ('+3', 'DDTRHIK'), ('-1', 'ILYA-CHP'), ('-2', 'FYMPSVI'), ('-3', 'FICLVSS')]\n",
      "splitsies ['WMTLGI', 'N'] 2\n",
      "outdic {'seq3': ['WMTLGI'], 'seq2': ['WITLGI'], 'seq1': ['MDYTRH'], 'seq5': ['MEHY'], 'seq4': ['WMTLGI', 'N']}\n"
     ]
    }
   ],
   "source": [
    "### FOR DEV ###\n",
    "print \"frm\", frm\n",
    "print\"frames\", frames\n",
    "print \"fastadic\", fastadic\n",
    "print \"trns\", trns\n",
    "print \"transzip\", transzip\n",
    "print \"splitsies\", splitsies, len(splitsies)\n",
    "print \"outdic\", outdic\n",
    "### END DEV ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">seq3_frame=+1;ORF_num=1\n",
      "WMTLGI\n",
      ">seq3_frame=+2;ORF_num=1\n",
      "G\n",
      ">seq3_frame=+2;ORF_num=2\n",
      "H\n",
      ">seq3_frame=+2;ORF_num=3\n",
      "AY\n",
      ">seq3_frame=+3;ORF_num=1\n",
      "DDTRHI\n",
      ">seq3_frame=-1;ORF_num=1\n",
      "LYA\n",
      ">seq3_frame=-1;ORF_num=2\n",
      "CHP\n",
      ">seq3_frame=-2;ORF_num=1\n",
      "YMPSVI\n",
      ">seq3_frame=-3;ORF_num=1\n",
      "ICLVSS\n",
      ">seq2_frame=+1;ORF_num=1\n",
      "WITLGI\n",
      ">seq2_frame=+2;ORF_num=1\n",
      "GLH\n",
      ">seq2_frame=+2;ORF_num=2\n",
      "AY\n",
      ">seq2_frame=+3;ORF_num=1\n",
      "DYTRHI\n",
      ">seq2_frame=-1;ORF_num=1\n",
      "LYA\n",
      ">seq2_frame=-1;ORF_num=2\n",
      "CNP\n",
      ">seq2_frame=-2;ORF_num=1\n",
      "YMPSVI\n",
      ">seq2_frame=-3;ORF_num=1\n",
      "ICLV\n",
      ">seq2_frame=-3;ORF_num=2\n",
      "S\n",
      ">seq1_frame=+1;ORF_num=1\n",
      "MDYTRH\n",
      ">seq1_frame=+2;ORF_num=1\n",
      "WITLGI\n",
      ">seq1_frame=+3;ORF_num=1\n",
      "GLH\n",
      ">seq1_frame=+3;ORF_num=2\n",
      "AL\n",
      ">seq1_frame=-1;ORF_num=1\n",
      "LMPSVIH\n",
      ">seq1_frame=-2;ORF_num=1\n",
      "CLV\n",
      ">seq1_frame=-2;ORF_num=2\n",
      "S\n",
      ">seq1_frame=-3;ORF_num=1\n",
      "NA\n",
      ">seq1_frame=-3;ORF_num=2\n",
      "CNP\n",
      ">seq5_frame=+1;ORF_num=1\n",
      "MEHY\n",
      ">seq5_frame=+2;ORF_num=1\n",
      "WNITNN\n",
      ">seq5_frame=+3;ORF_num=1\n",
      "GTLLII\n",
      ">seq5_frame=-1;ORF_num=1\n",
      "LLLVMFH\n",
      ">seq5_frame=-2;ORF_num=1\n",
      "YY\n",
      ">seq5_frame=-2;ORF_num=2\n",
      "CS\n",
      ">seq5_frame=-3;ORF_num=1\n",
      "IISNVP\n",
      ">seq4_frame=+1;ORF_num=1\n",
      "WMTLGI\n",
      ">seq4_frame=+1;ORF_num=2\n",
      "N\n",
      ">seq4_frame=+2;ORF_num=1\n",
      "G\n",
      ">seq4_frame=+2;ORF_num=2\n",
      "H\n",
      ">seq4_frame=+2;ORF_num=3\n",
      "AYK\n",
      ">seq4_frame=+3;ORF_num=1\n",
      "DDTRHIK\n",
      ">seq4_frame=-1;ORF_num=1\n",
      "ILYA\n",
      ">seq4_frame=-1;ORF_num=2\n",
      "CHP\n",
      ">seq4_frame=-2;ORF_num=1\n",
      "FYMPSVI\n",
      ">seq4_frame=-3;ORF_num=1\n",
      "FICLVSS\n"
     ]
    }
   ],
   "source": [
    "# write the translations to a file\n",
    "if outfile == 'stdout':\n",
    "    for nam, trns in outdic.items():\n",
    "        if seqbreak:\n",
    "            for frame in frames:\n",
    "                seqno = 0\n",
    "                for rec in trns[frame]:\n",
    "                    header = \"\".join([\">\", nam, \"_frame=\", frame, \";ORF_num=\", str(seqno + 1), \"\\n\"])\n",
    "                    print header, trns[frame][seqno]\n",
    "                    seqno += 1\n",
    "        else:\n",
    "            for frame in frames:\n",
    "                print \"\".join([\">\", nam, \"_frame=\", frame, \"\\n\", trns[frame]])\n",
    "else:\n",
    "    with open(outfile,\"w\") as finalout:\n",
    "        for nam, trns in outdic.items():\n",
    "            if seqbreak:\n",
    "                for frame in frames:\n",
    "                    seqno = 0\n",
    "                    for rec in trns[frame]:\n",
    "                        header = \"\".join([\">\", nam, \"_frame=\", frame, \";ORF_num=\", str(seqno + 1), \"\\n\"])\n",
    "                        print >> finalout, header, trns[frame][seqno]\n",
    "                        seqno += 1\n",
    "            else:\n",
    "                for frame in frames:\n",
    "                    print >> finalout, \"\".join([\">\", nam, \"_frame=\", frame, \"\\n\", trns[frame]])\n",
    "\n",
    "### EOF ###"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
