{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "### @package dna2aa \n",
    "##   Translate the input sequence in the specified or all frames.\n",
    "##\n",
    "##   Output: .fa file of translated subject sequences.\n",
    "##   These sequences are naively translated as internal ORFs,\n",
    "##   and are likely larger than the sequence's true ORF.\n",
    "##   The output is best suited to use for hmmer-like alignments.\n",
    "##\n",
    "##   Alternatively, using the 'seqbreak' option, ORFs are broken\n",
    "##   at stop codons.\n",
    "##\n",
    "##   Written by Austin Hammond, GSC 2015\n",
    "\n",
    "import argparse\n",
    "import sys\n",
    "import csv\n",
    "\n",
    "import fastatools as ft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO\n",
    "## output ORF sequence itself in addition to peptide\n",
    "## add eval of complete/partial ORF + peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mtdna2aa(seq):\n",
    "    tbl = {\"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"M\", \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "        \"TTA\":\"L\", \"TTG\":\"L\", \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\", \"TTT\":\"F\",\n",
    "        \"TTC\":\"F\", \"ATG\":\"M\", \"TGT\":\"C\", \"TGC\":\"C\", \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\",\n",
    "        \"GCG\":\"A\", \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\", \"CCT\":\"P\", \"CCC\":\"P\",\n",
    "        \"CCA\":\"P\", \"CCG\":\"P\", \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\", \"TCT\":\"S\",\n",
    "        \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\", \"TAT\":\"Y\", \"TAC\":\"Y\",\n",
    "        \"TGG\":\"W\", \"CAA\":\"Q\", \"CAG\":\"Q\", \"AAT\":\"N\", \"AAC\":\"N\", \"CAT\":\"H\", \"CAC\":\"H\",\n",
    "        \"GAA\":\"E\", \"GAG\":\"E\", \"GAT\":\"D\", \"GAC\":\"D\", \"AAA\":\"K\", \"AAG\":\"K\", \"CGT\":\"R\",\n",
    "        \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\", \"AGA\":\"-\", \"AGG\":\"-\", \"TAA\":\"-\", \"TAG\":\"-\",\n",
    "        \"TGA\":\"W\"}\n",
    "    j = 0 # for iteration\n",
    "    aa = \"\"\n",
    "    seql = len(seq)\n",
    "    if seql % 3 != 0:\n",
    "        seq = seq[0:(seql - seql%3)] #trim to multiple of 3, but should be already\n",
    "    seql = len(seq) #recalculate\n",
    "    for j in range(0, seql, 3):\n",
    "        j += 1\n",
    "        if j == (seql - 1):\n",
    "            codon = seq[j:]\n",
    "        else:\n",
    "            codon = seq[j-1:j+2]\n",
    "    # print codon\n",
    "        aa = aa + tbl[codon]\n",
    "    # force first codon to be M\n",
    "    aa = \"M\" + aa[1:]\n",
    "\n",
    "    return aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mtsixframe(seq):\n",
    "    one = mtdna2aa(seq)\n",
    "    two = mtdna2aa(seq[1:])\n",
    "    three = mtdna2aa(seq[2:])\n",
    "    none = mtdna2aa(ft.revcomp(seq))\n",
    "    ntwo = mtdna2aa(ft.revcomp(seq)[1:])\n",
    "    nthree = mtdna2aa(ft.revcomp(seq)[2:])\n",
    "\n",
    "    return [one, two, three, none, ntwo, nthree]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_met(sqn):\n",
    "    # trim the given sequence to the first methionine (M) residue\n",
    "    # if no M, then 'no_met' is returned\n",
    "    for pos in range(0, len(sqn)):\n",
    "        res = sqn[pos]\n",
    "        if res is not \"M\":\n",
    "            continue\n",
    "        else:\n",
    "            outsqn = sqn[pos:]\n",
    "            return outsqn\n",
    "    else:\n",
    "        return \"no_met\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_off_stop(sqn):\n",
    "    # trim off the stop codon marks that I introduce (-)\n",
    "    # only trims \"-\" at end of the peptide\n",
    "    if \"-\" not in sqn:\n",
    "        return sqn\n",
    "    rev = sqn[::-1]\n",
    "    for pos in range(0, len(rev)):\n",
    "        res = rev[pos]\n",
    "        if res is \"-\":\n",
    "            continue\n",
    "        else:\n",
    "            trim = rev[pos:][::-1]\n",
    "            break\n",
    "    return trim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "def len_thresh(sqn, threshold):\n",
    "    # drop sequences under a given length\n",
    "    # sequences where len >= threshold will be output\n",
    "    # sqn is a single translated sequence [STR]\n",
    "    # threshold = [INT]\n",
    "    # if len(sqn) < threshold, 'short_pep_LEN_residues' returned\n",
    "    thresh = int(threshold) # just in case...\n",
    "    sqnS = sqn.strip() # avoid trailing whitespace, newlines\n",
    "    sqnL = len(sqnS)\n",
    "    if sqnL >= threshold:\n",
    "        return sqnS\n",
    "    else:\n",
    "        msg = \"\".join([\"short_pep_\", str(sqnL), \"_residues\"])\n",
    "        return msg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup parser\n",
    "parser = argparse.ArgumentParser(\n",
    "    description=\"Naively translate the input DNA sequence.\")\n",
    "parser.add_argument(\n",
    "    \"--sequence\", \"-s\", action=\"store\", help=\"The fasta file of DNA sequences [STDIN]\",\n",
    "    default=\"stdin\")\n",
    "parser.add_argument(\"--frame\", \"-f\", action=\"store\",\n",
    "    choices=[\"+1\",\"+2\",\"+3\",\"-1\",\"-2\",\"-3\",\"all\"],\n",
    "    help=\"Frame to use for sequence translation [all]\", default=\"all\")\n",
    "parser.add_argument(\"--seqbreak\",\"-b\",action=\"store_true\",\n",
    "    help=\"Break translations at in-frame stop codons\")\n",
    "parser.add_argument(\"--outfile\", \"-o\", action=\"store\",\n",
    "    help=\"Name to use for output fasta file [STDOUT]\", default=\"stdout\")\n",
    "parser.add_argument(\"--mito\", \"-m\", action=\"store_true\", \n",
    "    help=(\"Sequence is mitochondrial; use vertebrate Mt code.\"\n",
    "          \" Does not allow stop codons to be completed by polyadenylation.\"))\n",
    "parser.add_argument(\"--min_length\", \"-l\", help=\"Minimum peptide length to output [30]\",\n",
    "                    action=\"store\", type=int, default=30)\n",
    "parser.add_argument(\"--trim_to_met\", \"-t\", help=\"Trim peptides to first M\",\n",
    "                    action=\"store_true\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "### FOR DEV ###\n",
    "frm = \"all\"\n",
    "outfile = \"stdout\"\n",
    "#outfile = \"tests/test-out.txt\"\n",
    "fasta = \"tests/untitled.txt\"\n",
    "fastadic = {}\n",
    "outdic = {}\n",
    "seqbreak = False\n",
    "mito = False\n",
    "#frm = \"+1\"\n",
    "trim = True\n",
    "minlen=5\n",
    "### END DEV ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse args\n",
    "args = parser.parse_args()\n",
    "frm = args.frame\n",
    "fasta = args.sequence\n",
    "fastadic = {}\n",
    "outfile = args.outfile\n",
    "outdic = {}\n",
    "mito = args.mito\n",
    "seqbreak = args.seqbreak\n",
    "minlen = args.min_length\n",
    "trim = args.trim_to_met"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check that the frame was passed properly\n",
    "if frm == \"all\":\n",
    "    pass\n",
    "elif \"+\" in frm:\n",
    "    pass\n",
    "elif frm == \"1\":\n",
    "    frm = \"+1\"\n",
    "elif frm == \"2\":\n",
    "    frm = \"+2\"\n",
    "elif frm == \"3\":\n",
    "    frm = \"+3\"\n",
    "elif \"-\" not in frm:\n",
    "    print \"ERROR: I didn't understand what reading frame you want.\"\n",
    "    print \"Please specify 'all' or a single frame including + or - and retry.\"\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define possible reading frames\n",
    "if frm == \"all\":\n",
    "    frames = [\"+1\",\"+2\",\"+3\",\"-1\",\"-2\",\"-3\"]\n",
    "else:\n",
    "    frames = [frm]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in sequence input\n",
    "if fasta == \"stdin\":\n",
    "    for rec in ft.fasta_iter(sys.stdin):\n",
    "        fastadic[rec[0].split(\" \")[0]] = rec[1]\n",
    "\n",
    "else:\n",
    "    with open(fasta) as fa:\n",
    "        for rec in ft.fasta_iter(fa):\n",
    "            sid = rec[0].split(\" \")[0]\n",
    "            sqn = rec[1]\n",
    "            fastadic[sid] = sqn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "# perform translations\n",
    "if seqbreak:\n",
    "    # setup separate dict to hold split translations\n",
    "    transdic = {}\n",
    "\n",
    "for sid, sqn in fastadic.items():\n",
    "    if mito == True:\n",
    "        if frm == \"all\":\n",
    "            trns = mtsixframe(sqn)\n",
    "        elif frm == \"+1\":\n",
    "            trns = mtdna2aa(sqn)\n",
    "        elif frm == \"+2\":\n",
    "            trns = mtdna2aa(sqn[1:])\n",
    "        elif frm == \"+3\":\n",
    "            trns = mtdna2aa(sqn[2:])\n",
    "        elif frm == \"-1\":\n",
    "            trns = mtdna2aa(ft.revcomp(sqn))\n",
    "        elif frm == \"-2\":\n",
    "            trns = mtdna2aa(ft.revcomp(sqn)[1:])\n",
    "        else:\n",
    "            trns = mtdna2aa(ft.revcomp(sqn)[2:])\n",
    "\n",
    "    else:\n",
    "        if frm == \"all\":\n",
    "            trns = ft.sixframe(sqn)\n",
    "        elif frm == \"+1\":\n",
    "            trns = ft.dna2aa(sqn)\n",
    "        elif frm == \"+2\":\n",
    "            trns = ft.dna2aa(sqn[1:])\n",
    "        elif frm == \"+3\":\n",
    "            trns = ft.dna2aa(sqn[2:])\n",
    "        elif frm == \"-1\":\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn))\n",
    "        elif frm == \"-2\":\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn)[1:])\n",
    "        else:\n",
    "            trns = ft.dna2aa(ft.revcomp(sqn)[2:])\n",
    "\n",
    "    # setup output dict entry\n",
    "    outdic[sid] = {}\n",
    "    if seqbreak:\n",
    "        # if frm == all, trns is a list; otherwise is a string #TODO make always a list\n",
    "        if type(trns) == list:\n",
    "            if frm == \"all\":\n",
    "                transzip = zip(frames, trns)\n",
    "            else:\n",
    "                transzip = (frm, trns)\n",
    "            for rec in transzip:\n",
    "                thisframe = rec[0]\n",
    "                thisseq = rec[1]\n",
    "                #splitsies = [x.split(\"-\") for x in thisseq]\n",
    "                splitsies = thisseq.split(\"-\")\n",
    "                # drop empty frames\n",
    "                splitsies = [x for x in splitsies if x]\n",
    "                outdic[sid][thisframe] = splitsies\n",
    "        else:\n",
    "            splitsies = trns.split(\"-\") # split into ORFs\n",
    "            multi = [x for x in splitsies if x] # drop empty ORFs\n",
    "            outdic[sid][frm] = multi\n",
    "    else:\n",
    "        if type(trns) == list:\n",
    "            outdic[sid] = {}\n",
    "            if frm == \"all\":\n",
    "                transzip = zip(frames, trns)\n",
    "            else:\n",
    "                transzip = (frm, trns)\n",
    "            for rec in transzip:\n",
    "                thisframe = rec[0]\n",
    "                thisseq = rec[1]\n",
    "                outdic[sid][thisframe] = thisseq\n",
    "        else:\n",
    "            outdic[sid][frm] = trns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "frm all\n",
      "frames ['+1', '+2', '+3', '-1', '-2', '-3']\n",
      "fastadic {'seq3': 'TGGATGACACTAGGCATATAA', 'seq2': 'TGGATTACACTAGGCATATAA', 'seq1': 'ATGGATTACACTAGGCATTAA', 'seq5': 'ATGGAACATTACTAATAATAA', 'seq4': 'TGGATGACACTAGGCATATAAAAT'}\n",
      "trns ['WMTLGI-N', 'G-H-AYK', 'DDTRHIK', 'ILYA-CHP', 'FYMPSVI', 'FICLVSS']\n",
      "transzip [('+1', 'WMTLGI-N'), ('+2', 'G-H-AYK'), ('+3', 'DDTRHIK'), ('-1', 'ILYA-CHP'), ('-2', 'FYMPSVI'), ('-3', 'FICLVSS')]\n",
      "splitsies ['FICLVSS'] 1\n",
      "outdic {'seq3': {'+3': 'DDTRHI', '+2': 'G-H-AY', '+1': 'WMTLGI-', '-1': 'LYA-CHP', '-3': 'ICLVSS', '-2': 'YMPSVI'}, 'seq2': {'+3': 'DYTRHI', '+2': 'GLH-AY', '+1': 'WITLGI-', '-1': 'LYA-CNP', '-3': 'ICLV-S', '-2': 'YMPSVI'}, 'seq1': {'+3': 'GLH-AL', '+2': 'WITLGI', '+1': 'MDYTRH-', '-1': 'LMPSVIH', '-3': 'NA-CNP', '-2': '-CLV-S'}, 'seq5': {'+3': 'GTLLII', '+2': 'WNITNN', '+1': 'MEHY---', '-1': 'LLLVMFH', '-3': 'IISNVP', '-2': 'YY--CS'}, 'seq4': {'+3': 'DDTRHIK', '+2': 'G-H-AYK', '+1': 'WMTLGI-N', '-1': 'ILYA-CHP', '-3': 'FICLVSS', '-2': 'FYMPSVI'}}\n",
      "no_met\n",
      "MPSVI\n"
     ]
    }
   ],
   "source": [
    "### FOR DEV ###\n",
    "print \"frm\", frm\n",
    "print\"frames\", frames\n",
    "print \"fastadic\", fastadic\n",
    "print \"trns\", trns\n",
    "print \"transzip\", transzip\n",
    "print \"splitsies\", splitsies, len(splitsies)\n",
    "print \"outdic\", outdic\n",
    "print trim_met(outdic[\"seq2\"][\"+1\"])\n",
    "print len_thresh(trim_met(transzip[4][1]), 2)\n",
    "### END DEV ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">seq3_frame=+1\n",
      "MTLGI\n",
      ">seq3_frame=-2\n",
      "MPSVI\n",
      ">seq2_frame=-2\n",
      "MPSVI\n",
      ">seq1_frame=+1\n",
      "MDYTRH\n",
      ">seq1_frame=-1\n",
      "MPSVIH\n",
      ">seq4_frame=+1\n",
      "MTLGI-N\n",
      ">seq4_frame=-2\n",
      "MPSVI\n"
     ]
    }
   ],
   "source": [
    "# write the translations to a file\n",
    "# TODO clean up this redundant code\n",
    "if outfile == \"stdout\":\n",
    "    for nam, trns in outdic.items():\n",
    "        if seqbreak:\n",
    "            for frame in frames:\n",
    "                seqno = 0\n",
    "                for rec in trns[frame]:\n",
    "                    # get rid of trailing \"-\"\n",
    "                    pep = trim_off_stop(trns[frame][seqno])\n",
    "                    # apply trim_met before length cutoff\n",
    "                    if trim:\n",
    "                        trimmed = trim_met(pep)\n",
    "                        if \"no_met\" in trimmed:\n",
    "                            continue\n",
    "                        else:\n",
    "                            # check peptide length >= minlen\n",
    "                            if \"short_pep\" not in len_thresh(trimmed, minlen):\n",
    "                                header = \"\".join([\">\", nam, \"_frame=\", frame,\n",
    "                                                  \";ORF_num=\", str(seqno + 1),\n",
    "                                                  \"\\n\"])\n",
    "                                print header, trimmed\n",
    "                                seqno += 1\n",
    "                            else:\n",
    "                                continue\n",
    "                    else:\n",
    "                        # check peptide length >= minlen\n",
    "                        if \"short_pep\" not in len_thresh(pep, minlen):\n",
    "                            header = \"\".join([\">\", nam, \"_frame=\", frame,\n",
    "                                              \";ORF_num=\", str(seqno + 1),\n",
    "                                              \"\\n\"])\n",
    "                            print header, pep\n",
    "                            seqno += 1\n",
    "                        else:\n",
    "                            continue\n",
    "        else:\n",
    "            for frame in frames:\n",
    "                # get rid of trailing \"-\"\n",
    "                pep = trim_off_stop(trns[frame])\n",
    "                if trim:\n",
    "                    trimmed = trim_met(pep)\n",
    "                    if \"no_met\" in trimmed:\n",
    "                        continue\n",
    "                    else:\n",
    "                        if \"short_pep\" not in len_thresh(trimmed, minlen):\n",
    "                            print \"\".join([\">\", nam, \"_frame=\", frame, \"\\n\", trimmed])\n",
    "                        else:\n",
    "                            continue\n",
    "                else:\n",
    "                    if \"short_pep\" not in len_thresh(pep, minlen):\n",
    "                        print \"\".join([\">\", nam, \"_frame=\", frame, \"\\n\", pep])\n",
    "                    else:\n",
    "                        continue\n",
    "else:\n",
    "    with open(outfile,\"w\") as finalout:\n",
    "        for nam, trns in outdic.items():\n",
    "            if seqbreak:\n",
    "                for frame in frames:\n",
    "                    seqno = 0\n",
    "                    for rec in trns[frame]:\n",
    "                        pep = trim_off_stop(trns[frame][seqno])\n",
    "                        if trim:\n",
    "                            trimmed = trim_met(pep)\n",
    "                            if \"no_met\" in trimmed:\n",
    "                                continue\n",
    "                            else:\n",
    "                                # check peptide length >= minlen\n",
    "                                if \"short_pep\" not in len_thresh(trimmed, minlen):\n",
    "                                    header = \"\".join([\">\", nam, \"_frame=\", frame,\n",
    "                                                      \";ORF_num=\", str(seqno + 1),\n",
    "                                                      \"\\n\"])\n",
    "                                    print header, trimmed\n",
    "                                    seqno += 1\n",
    "                                else:\n",
    "                                    continue\n",
    "                        else:\n",
    "                            # check peptide length >= minlen\n",
    "                            if \"short_pep\" not in len_thresh(pep, minlen):\n",
    "                                header = \"\".join([\">\", nam, \"_frame=\", frame,\n",
    "                                                  \";ORF_num=\", str(seqno + 1), \"\\n\"])\n",
    "                                print >> finalout, header, pep\n",
    "                                seqno += 1\n",
    "                            else:\n",
    "                                continue\n",
    "            else:\n",
    "                for frame in frames:\n",
    "                    pep = trim_off_stop(trns[frame][seqno])\n",
    "                    if trim:\n",
    "                        trimmed = trim_met(pep)\n",
    "                        if \"no_met\" in trimmed:\n",
    "                            continue\n",
    "                        else:\n",
    "                            # check peptide length >= minlen\n",
    "                            if \"short_pep\" not in len_thresh(trimmed, minlen):\n",
    "                                print >> finalout, \"\".join([\">\", nam, \"_frame=\",\n",
    "                                                            frame, \"\\n\", trimmed])\n",
    "                            else:\n",
    "                                continue\n",
    "                    else:\n",
    "                        # check peptide length >= minlen\n",
    "                        if \"short_pep\" not in len_thresh(pep, minlen):\n",
    "                            print >> finalout, \"\".join([\">\", nam, \"_frame=\", frame,\n",
    "                                                        \"\\n\", pep])\n",
    "                        else:\n",
    "                            continue\n",
    "\n",
    "### EOF ###"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
